jupyterlab/webpack.config.js

```diff
@@ -6,21 +6,193 @@
+        }
+
+        // Look for loaders
+        if (module.loaders.length) {
+          modRequest = modRequest.slice(modRequest.lastIndexOf('!') + 1);
+          return;
+        }
+
+        var source = module.source().source();
+        for (var dep of deps) {
+          var id = dep.id;
+          var request = dep.request;
+          var search = '__webpack_require__(' + id + ')';
+          if (!request) {
+            // TODO: better discrimination of these.
+            source = source.split(search).join('UNDEFINED');
+            continue;
+          }
+          if (request.indexOf('extract-text-webpack-plugin') !== -1) {
+            source = source.replace(search + ';\n', '');
+            continue;
+          }
+          if (dep.loaders.length) {
+            request = request.slice(request.lastIndexOf('!') + 1);
+          }
+          request = findImport(request, dep.issuer);
+          request = 'require("jupyterlab!' + request + '")';
+          source = source.split(search).join(request);
+        }
+        var modPath = findName(modRequest);
+        modules.push(modPath);
+        var header = 'define("' + modPath;
+        header += '", function (require, exports, module) {\n'
+        // Combine with indent.
+        source = header + source.split('\n').join('\n  ') + '\n})'
+        sources.push(source);
+      });
+    });
+
+    // Assemble the r.js bundle.
+    // TODO: Add config for dynamic modules.
+    // var code = 'require.config({ baseUrl: "'
+    //code += compilation.outputOptions.publicPath + '" }),\n\n'
+    //code += sources.join(',\n\n');
+
+    var code = sources.join(',\n\n');
+
+    // Insert this list into the Webpack build as a new file asset:
+    compilation.assets['custom.bundle.js'] = {
+      source: function() {
+        return code;
+      },
+      size: function() {
+        return code.length;
+      }
+    };
+
+    compilation.assets['custom.manifest.txt'] = {
+      source: function() {
+        return modules.join('\n');
+      },
+      size: function() {
+        return modules.join('\n').length;
+      }
+    }
+    callback();
+  });
+};
+
+
+// From a request - find its package root.
+function findRoot(request) {
+  if (path.extname(request)) {
+    request = path.dirname(request);
+  }
+  while (true) {
+    try {
+      require.resolve(path.join(request, 'package.json'));
+      return request;
+    } catch (err) {
+      request = path.dirname(request);
+    }
+  }
+}
 
-console.log('Generating bundles...');
 
-module.exports = {
+// Get the package.json associated with a file.
+function getPackage(request) {
+  var rootPath = findRoot(request);
+  return require(path.join(rootPath, 'package.json'));
+}
+
+
+// From a request - find a version-mangled define name.
+function findName(request) {
+  var rootPath = findRoot(request);
+  var package = getPackage(rootPath);
+  var modPath = request.slice(rootPath.length + 1);
+  var name = package.name + ':' + package.version;
```

Comment:
I appreciate that. It's not a public API yet, but it is an abi issue... Changing this means people have to recompile their extensions and make a new release. When we support dynamic loading, it will be a public API (or naturally could be).